File: app/src/main/java/com/example/uos_lms/core/domain/model/Message.java
package com.example.uos_lms.core.domain.model;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@Getter
@Builder(toBuilder = true)
@EqualsAndHashCode
@ToString
public class Message {
    @Builder.Default
    private final String id = "";
    @Builder.Default
    private final String conversationId = "";
    @Builder.Default
    private final String senderUid = "";
    @Builder.Default
    private final String senderName = "";
    @Builder.Default
    private final String senderRole = "";
    @Builder.Default
    private final String recipientUid = "";
    @Builder.Default
    private final String text = "";
    private final String attachmentUrl;
    private final String attachmentName;
    private final String attachmentPublicId;
    private final String attachmentResourceType;
    private final Long attachmentSize;
    @Builder.Default
    private final long sentAt = 0L;
    /** Null = not yet delivered/read - the recipient's client hasn't fetched (deliveredAt) or
     * opened the thread for (readAt) this message. Drives the single/double/blue-tick UI. */
    private final Long deliveredAt;
    private final Long readAt;
}

File: src/models/Message.js
const mongoose = require('mongoose');
const { Schema } = mongoose;

const messageSchema = new Schema(
  {
    conversationId: { type: String, ref: 'Conversation', required: true },
    senderId: { type: String, ref: 'User', required: true },
    recipientId: { type: String, ref: 'User', required: true },
    text: { type: String, default: '' },

    mediaId: { type: Schema.Types.ObjectId, ref: 'Media', default: null },
    attachmentUrl: { type: String, default: null },
    attachmentName: { type: String, default: null },
    attachmentPublicId: { type: String, default: null },
    attachmentResourceType: { type: String, default: null },
    attachmentSize: { type: Number, default: null },

    sentAt: { type: Date, default: Date.now },
    // Stamped when the recipient's client fetches this message at all (listMessages) and
    // when they actually open the thread (markRead) - single-tick/double-tick/blue-tick
    // semantics without a real-time channel, since this backend is REST/polling only.
    deliveredAt: { type: Date, default: null },
    readAt: { type: Date, default: null },
  },
  { timestamps: true }
);

messageSchema.index({ conversationId: 1, sentAt: 1 });

module.exports = mongoose.model('Message', messageSchema);
